[Board] board domain 계층 구현

vwx12345 opened the pull request:
### 💻 작업 내용
- Board 도메인 영역 구성 (Entity, Value Object, Domain Service, Exception 정의)

### ✨ 리뷰 포인트
- BoardNameUniquenessCheck를 통해서 중복성 체크를 구현했습니다.

### 📝 메모


### 🎯 관련 이슈
closed #86

vwx12345 committed:
feat(Board): Board Aggregate-root 구현

- Visitor로 받아서 Manager로 전환하는 방식
- 처음에는 create로 boardId를 null로 받고 이후 jpa 저장 이후 boardId를 지정해주는 방식으로 구현

vwx12345 committed:
feat(Board): BoardName Vo 구현

vwx12345 committed:
feat(Board): Manager Vo 구현

vwx12345 committed:
feat(Board): BoardNameUniquenessChecker 구현

- BoardNameUniquenessChecker 구현
- DuplicateBoardNameException 정의

vwx12345 committed:
feat(Board): InvalidBoardNameException 정의

vwx12345 committed:
feat(Board): BoardRepository 정의

vwx12345 committed:
docs(Board): BoardNameUniquenessChecker 주석 추가

vwx12345 committed:
test(Board): BoardNameTest 추가

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/service/BoardNameUniquenessChecker.java:
```diff
@@ -0,0 +1,14 @@
+
+  default void ensureUnique(BoardName name) {
+    if (!isUnique(name)) {
+      throw new DuplicateBoardNameException("이미 존재하는 보드명: " + name);
```
isUnique 메서드의 경우 기능이 어느정도 예상이 가는데 ensureUnique는 정확히 어떤 메서드인지 모르겠네요.
둘은 어떤 차이가 있나요?
isUnique의 경우에는 워낙 이름이 투명하니 주석이 없어도 괜찮은데, ensureUnique 메서드의 경우에는 주석이 필요해보이네요.
default로 되어있는데, ensureUnique 메서드의 기능과 책임이 뭔지, 오버라이딩이 된다면 어떤식으로 될지? 이런게 궁금합니다.
default로 선언되어있는 이유도 궁금하네요.

vwx12345 replied:
isUnique는 단순히 보드명이 유일한지 여부를 boolean으로 리턴 메서드, ensureUnique는 “유일하지 않다면 예외를 던진다”는 책임을 캡슐화한 메서드라고 보면 될 것 같아요. 하지만 확실히 ensureUnique만 보면 동작을 모를 수 있겠네요. 주석 추가하겠습니다.

인터페이스에서 default를 쓰면, 모든 구현체가 공통적으로 쓸 수 있는 기본 구현을 제공할 수 있습니다. 여기서는 isUnique만 구현하면 ensureUnique는 자동으로 쓸 수 있도록 구현한 것입니다. 만약 특정 구현체가 다른 예외 처리 방식이나 로깅을 추가하고 싶다면, ensureUnique를 오버라이딩해서 자신만의 동작을 넣을 수도 있습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/service/BoardNameUniquenessChecker.java:
```diff
@@ -3,6 +3,14 @@
+ *
+ * <p>보드명 중복 여부를 확인하는 도메인 서비스
+ *
+ * @domain domain-service
```
분류도 도메인 서비스로 되어있고, 주석도 도메인 서비스로 달려있다면 클래스 명에도 도메인 서비스가 들어가있어야 합니다.
"Checker"라는 명칭은 Validator 혹은 Utility 클래스와 헷갈릴것같아요.

vwx12345 replied:
확실히 유틸리티랑 헷갈릴 수 있을 것 같네요. 다시 네이밍해서 pr하겠습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/service/BoardNameUniquenessChecker.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface BoardNameUniquenessChecker {
+  boolean isUnique(BoardName name);
+
```
BoardNameUniquenessChecker 인터페이스가 정의된 커밋 제목이
"feat(Board): BoardNameUniquenessChecker 구현" 인데, 인터페이스 정의밖에 없네요.
이런 경우는 "인터페이스 정의"가 커밋 제목에 더 어울리지 않을까 합니다.

또, 주석을 보니 "BoardNameUniquenessChecker"는 도메인 서비스 클래스인데 도메인 계층 개발 PR에서 구현체까지 개발하지 않고 인터페이스만 정의하신 이유가 있으신가요?

vwx12345 replied:
정의가 맞는 것 같습니다. 다른건 정의라고 했는데 이 부부은 실수했네요.

아래에 대한 대답은 저는 "보드명은 유일해야한다"라는 도메인 규칙을 정의한 것이고 그렇기 때문에 실제 구현은 다른 계층에서 이루어질 예정입니다.

codingbaraGo replied:
> 정의가 맞는 것 같습니다. 다른건 정의라고 했는데 이 부부은 실수했네요.
> 
> 아래에 대한 대답은 저는 "보드명은 유일해야한다"라는 도메인 규칙을 정의한 것이고 그렇기 때문에 실제 구현은 다른 계층에서 이루어질 예정입니다.

음... 저는 도메인 규칙 전체를 추상화하기 보다는 도메인 규칙은 도메인 영역에 작성하고, 인프라에 의존성이 있는 부분(BoardName을 DB에 조회하는 레포지토리 혹은 Mapper 등)만 인터페이스로 추상화하는게 좋지 않을까 생각했습니다.

근데 관련 기능이 DB에 의존성이 크다보니 저런 구조를 사용하는것도 가능하겠다 싶네요.
무엇이 더 좋은지는 잘 모르겠네요.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/vo/BoardName.java:
```diff
@@ -0,0 +1,49 @@
+    if (trimmed.length() < 2 || trimmed.length() > 30) {
+      throw new InvalidBoardNameException("보드명은 2~30자여야 합니다.");
+    }
+  }
```
Validation에 대한 정책은 "각 도메인 엔티티에서 맡는다" 로 정하신건가요?
Board B.C의 경우 공동 작업 B.C이니 정책을 통일해야 할것같네요.
별도의 Validator 클래스를 사용하지 않고 각 도메인 엔티티가 자기 컬럼에 대한 Validation을 수행하는 방향으로 정하신건가요?

codingbaraGo replied:
전 아직 Validation은 Presentation 계층 Validation정도만 구현중이고, 나머지 파트에 대한 validation은 보류상태인데, 관련 정책에 대해 정한게 있다면 노션에 문서를 하나 작성해주시면 참고해서 개발하겠습니다.
계층별 DTO 중 어떤 계층의 DTO들에서 Validation을 수행할지, converter, mapper에서는 어떻게 할지, 각 도메인 엔티티에서 validation을 수행한다면, 따로 infra entity에서는 validation을 수행하지 않을지 등에 관해 결정하신 사안이 있다면 기록 부탁드려요.

vwx12345 replied:
> Validation에 대한 정책은 "각 도메인 엔티티에서 맡는다" 로 정하신건가요? Board B.C의 경우 공동 작업 B.C이니 정책을 통일해야 할것같네요. 별도의 Validator 클래스를 사용하지 않고 각 도메인 엔티티가 자기 컬럼에 대한 Validation을 수행하는 방향으로 정하신건가요?

이 부분은 그때 구두로 얘기했던 것처럼 presentation 계층에서 검증하는 걸로 하고 삭제하도록 하겠습니다.

vwx12345 replied:
> 전 아직 Validation은 Presentation 계층 Validation정도만 구현중이고, 나머지 파트에 대한 validation은 보류상태인데, 관련 정책에 대해 정한게 있다면 노션에 문서를 하나 작성해주시면 참고해서 개발하겠습니다. 계층별 DTO 중 어떤 계층의 DTO들에서 Validation을 수행할지, converter, mapper에서는 어떻게 할지, 각 도메인 엔티티에서 validation을 수행한다면, 따로 infra entity에서는 validation을 수행하지 않을지 등에 관해 결정하신 사안이 있다면 기록 부탁드려요.

이 부분은 제가 혼자 결정할 부분은 아닌 것 같아서 회의 때 같이 얘기해보는 걸로 하죠.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/repository/BoardRepository.java:
```diff
@@ -0,0 +1,15 @@
+import com.simpleboard.board.board.domain.board.entity.Board;
+import java.util.Optional;
+
+public interface BoardRepository {
+
+  Board save(Board board);
```
Board aggregate의 경우 Command와 Query 레포를 구분해야하지 않을까요?

vwx12345 replied:
저는 Query 레포에는 도메인 로직 없기 때문에 domain 계층에 둘 필요 없다고 생각합니다. 따라서 application 계층에 구현하려고 했습니다.

codingbaraGo replied:
> 저는 Query 레포에는 도메인 로직 없기 때문에 domain 계층에 둘 필요 없다고 생각합니다. 따라서 application 계층에 구현하려고 했습니다.

그러면 Command 레포지토리의 이름은 {aggregate 이름} + "Respository" 형태로 작성하고,
Query 레포지토리의 이름은 {aggregate 이름} + "Query" + "Repository"로 작성하시는건가요?

domain 계층에 command 레포지토리만 존재한다고 해도 그냥 Command를 붙이는 편이 낫지 않을까요?

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/entity/Board.java:
```diff
@@ -0,0 +1,49 @@
+  public static Board create(
+      BoardName boardName, Visitor visitor, BoardNameUniquenessChecker checker) {
+    checker.ensureUnique(boardName);
+    return new Board(null, boardName, Manager.of(visitor.memberId()), LocalDateTime.now());
```
Checker 클래스는 도메인 서비스 클래스로 아는데, argument로 가져오는 이유가 있을까요?

vwx12345 replied:
우선 코드가 복잡하지 않고 생성 시점에 반드시 검사를 해야하기 때문에 저렇게 구현했습니다.

codingbaraGo commented on src/main/java/com/simpleboard/board/board/domain/board/entity/Board.java:
```diff
@@ -0,0 +1,49 @@
+
+  public Board withId(Long id) {
+    return new Board(id, this.boardName, this.manager, this.createdAt);
+  }
```
withId 메서드는 용도가 뭔가요?
어떤 상황에 사용하는 메서드인가요?
static이 안붙어있으니 팩토리메서드는 아닌듯한데, 
언제 사용하는건지 적어주시면 좋을것같아요.

지금과 같이 생성 메서드가 여러개 있을 때, argument로 DTO를 받을 경우, DTO의 이름으로 해당 생성 메서드가 사용되는 상황을 유추할 수 있고, 실제로 해당 생성 메서드를 호출하는 상황도 제한을 할 수 있지만(해당 DTO가 사용될때== 특정 USECASE 에만 호출할 수 있으니까요.) 지금처럼 argument로 필드값들을 받는 경우 어떤 생성 메서드를 언제 사용할지 주석이나 별도의 정책 문서로 정의해주어야 유지/보수성이 좋아질 것 같아요.

vwx12345 replied:
withId 메서드는 불변 객체처럼 ID만 교체한 새로운 인스턴스를 만들 때 쓰는 용도로, DB에 저장한 뒤 생성된 식별자를 반영할 때 사용하기 위해 추가한 메서드입니다.

하지만 JPA 환경에서는 @GeneratedValue를 통해 영속화 시점에 ID가 자동 할당되므로, 실제로 사용할 일이 없습니다. 따라서 당장은 불필요하다고 판단해 삭제하겠습니다.

vwx12345 replied:
생성 메서드가 많아질수록 DTO 기반 생성보다 의도를 드러내는 메서드 명명이 더 낫다고 생각합니다. 이후 생성 경로가 복잡해지면 Factory 패턴 도입도 검토하겠습니다.

codingbaraGo replied:
> 생성 메서드가 많아질수록 DTO 기반 생성보다 의도를 드러내는 메서드 명명이 더 낫다고 생각합니다. 이후 생성 경로가 복잡해지면 Factory 패턴 도입도 검토하겠습니다.

Factory 패턴 관련 이야기는, withId가 생성자도 팩토리메서드도 아니어서 withId를 사용하려면 Board 인스턴스를 하나 생성한 뒤, 생성된 인스턴스를 통해 withId를 호출해야하기 때문에 언급하였습니다.

codingbaraGo reviewed (commented):
PR의 리뷰포인트인 BoardNameUniquenessChecker 중심으로 리뷰했습니다.

vwx12345 committed:
refactor(Board): 리뷰 반영

- Board 의 사용하지 않는 생성 method withId 제거
- BoardName 에서 기본적인 검증 제거
- BoardNameUniquenessChecker에서 BoardNameUniquenessService로 이름 변경
- BoardNameUniquenessService 에서 ensureUnique 메서드 주석 추가

codingbaraGo reviewed (approved):
굿입니다.
한가지 제안은 BoardNameUniquenessService의 클래스명으로 BoardNamePrincipal은 어떤지 입니다.
도메인 서비스 클래스보다는 도메인 정책에 가까운 것 같아서요.
본인이 더 적합하다고 생각하시는 클래스 명을 사용하시면 될 것 같습니다.

vwx12345 committed:
refactor(board): BoardNameUniquenessService 이름 변경

- BoardNameUniquenessService 에서 BoardNamePrincipal 로 이름 변경

vwx12345 commented:
> 굿입니다. 한가지 제안은 BoardNameUniquenessService의 클래스명으로 BoardNamePrincipal은 어떤지 입니다. 도메인 서비스 클래스보다는 도메인 정책에 가까운 것 같아서요. 본인이 더 적합하다고 생각하시는 클래스 명을 사용하시면 될 것 같습니다.

저도 그게 더 좋을 것 같네요. 수정했습니다.
